File: templates/python/roc.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""ROC 曲线 + AUC —— 卡片 06_预测模型评价图/ROC曲线_ROC.md
合成预测 + sklearn。改 # >>> PARAM 区换 y_true / y_score。"""
import os, sys, warnings; warnings.filterwarnings("ignore")
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import numpy as np, matplotlib.pyplot as plt
from sklearn.metrics import roc_curve, roc_auc_score
from _common import set_theme, save_fig, pal
set_theme(); np.random.seed(7)

# >>> PARAM ------------------------------------------------------
NAME = "tpl_roc"
TITLE = "ROC curve"; XLAB = "1 − Specificity"; YLAB = "Sensitivity"
CURVE_COLOR = pal("med_case_control")["low"]
FIGSIZE = (3.2, 3.2)
# 换真实数据： y_true=df["y"]; y_score=model.predict_proba(X)[:,1]
n = 500
y_true = np.random.binomial(1, .4, n)
y_score = np.clip(.5 * y_true + np.random.normal(0, .4, n), 0, 1)
# <<< PARAM ------------------------------------------------------

fpr, tpr, _ = roc_curve(y_true, y_score); auc = roc_auc_score(y_true, y_score)
fig, ax = plt.subplots(figsize=FIGSIZE)
ax.plot(fpr, tpr, color=CURVE_COLOR, lw=1.8)
ax.plot([0, 1], [0, 1], ls="--", c="grey", lw=.8)
ax.text(.45, .12, f"AUC = {auc:.2f}", fontsize=9)
ax.set_aspect("equal"); ax.grid(False)
ax.set_xlabel(XLAB); ax.set_ylabel(YLAB); ax.set_title(TITLE)
save_fig(fig, NAME)

File: templates/r/roc.R
## ROC 曲线 + AUC 95%CI（pROC + ggplot）—— 卡片 06_预测模型评价图/ROC曲线_ROC.md
## 合成预测，可独立运行： Rscript templates/r/roc.R
.args <- commandArgs(trailingOnly = FALSE)
.sd <- dirname(normalizePath(sub("^--file=", "", .args[grep("^--file=", .args)])))
source(file.path(.sd, "_common.R"))
suppressPackageStartupMessages(library(pROC))
set.seed(2026)

# >>> PARAM ------------------------------------------------------
NAME <- "tpl_roc"; TITLE <- "ROC curve"; XLAB <- "1 - Specificity"; YLAB <- "Sensitivity"
CURVE <- unname(pal("med_case_control")["low"])
# 换真实数据： y(0/1) 与 pred(预测概率)
n <- 500; y <- rbinom(n, 1, .4); pred <- pmin(pmax(.5 * y + rnorm(n, 0, .4), 0), 1)
# <<< PARAM ------------------------------------------------------

roc_obj <- roc(y, pred, quiet = TRUE); ci_auc <- ci.auc(roc_obj)
roc_df <- data.frame(spec = rev(roc_obj$specificities), sens = rev(roc_obj$sensitivities))
p <- ggplot(roc_df, aes(1 - spec, sens)) +
  geom_abline(slope = 1, intercept = 0, linetype = 2, color = "grey60") +
  geom_path(color = CURVE, linewidth = 1) +
  annotate("text", x = .62, y = .12, size = 2.8,
           label = sprintf("AUC = %.3f\n(95%% CI %.3f-%.3f)", ci_auc[2], ci_auc[1], ci_auc[3])) +
  coord_equal() + labs(x = XLAB, y = YLAB, title = TITLE)
save_gg(p, NAME, w = 88, h = 88)
